spec/rubocop/cop/rspec/unexpected_requires_spec.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+RSpec.describe RuboCop::Cop::RSpec::UnexpectedRequires do
+  context 'when there is an require' do
+    it 'registers an offense' do
+      expect_offense(<<~RUBY)
+        require "base64"
```

Comment:
I could find `require "rspec-parameterized"` and also "fast_spec_helper" in the same spec file. Both seem legitimate.

Reply:
shouldn't those requires then be inside spec_helper / rails_helper then?

Reply:
I ran it against the real-world-specs, and if i did it right, there were no offences with this rubocop, but plenty of others ;-)

Reply:
Is there an easy way to run rubocop against all repos in real-world-rspec repo?

Reply:
Should a "fast_spec_helper" be required inside spec_helper? No, because supposedly the fast spec helper is a [subset of spec_helper](https://github.com/gitlabhq/gitlabhq/blob/ead25aaac3ab8e8883af13ec2e3b5308cf61ef70/spec/fast_spec_helper.rb#L4) that is purposed to run faster than spec_helper. I don't know the details, though.

Should a "rspec-parameterized" be required inside spec_helper? If it would, this spec alone will require spec_helper to load. And any spec requiring spec_helper would automatically require rspec-parameterized even if it's not used by that spec.

Reply:
> an easy way to run rubocop against all repos in real-world-rspec repo?

I can't call it easy, but:
1. You clone it with submodules (this may require removing some of the dead ones or ones that changed the primary branch)
2. You recursively find .rubocop*.yml and remove it
3. You run the cop from rubocop-rspec with `rubocop --only <your-cop> ../real-world-rspec`

I kindly appreciate it if you could document this in https://docs.rubocop.org/rubocop/development.html#testing-your-cop-in-a-real-codebase

Reply:
I ran it against the real-world-rspecs:

Prepare real-world-rspecs folder:
```
git clone git@github.com:pirj/real-world-rspec.git
cd real-world-spec
git submodule update --init --remote --checkout --single-branch --depth 1
find ./ -type f -name .rubocop.yml -exec rm -rf {} \;
```

```
cd rubocop-rspec
bundle exec rubocop --only RSpec/UnexpectedRequires ../real-world-rspec/
```

It finds many occurences ... but mostly: 
* pry
* helper
* test/unit
* active_support/*
* simplecov
* webmock
* rspec/support/*
* support/*
* tmpdir

and others. Which might be "Ok" to require in specs, but still i would more expect them to be required in spec_helper than in the individual files. 

So this does not really change my mind, that it should be avoided to add requires to spec-files. And as you can disable rubocops anyways when necessary, it still should be fine. 

I also saw this: 

```
require File.dirname(__FILE__) + "/../../spec_helper"
```

which could still be excluded in the checks .. but imo i would prefer "require 'spec_helper'" there anyways, not using a relative file-path .. but that is a different topic.

Reply:
[This](https://github.com/rspec/rspec-core/blob/2ba0f10fe68948e5ee9addd569b4694f0245aa71/lib/rspec/core/project_initializer/spec/spec_helper.rb#L11) is interesting.

To me, however, it’s unclear if we should create a separate webmock_spec_helper.rb with a single line that just requires webmock.

> production … was not able to start up due to missing requires

And I’m uncertain how imposing such a structure will solve the original issue.

Also, I don’t quite understand the original issue. Isn’t Bundler supposed to load all requirements? Shouldn’t own code be eagerly autoloaded in production (assuming a Rails app)?

Please don’t get me wrong, I still recall a very-well covered feature that relied on a gem (json schema validator) that was only present in the `:test` and `:development` bundler groups, making it pass specs perfectly, but fail in production.
I can’t tell off the top of my head how `require: false` would affect this.

There are also non-Rails apps with, supposedly, explicit require structure. And non-eagerly autoloaded paths in Rails apps, too (support lib code for rake tasks; internal rubocop cops, …).

I have no certainty that just banning requires from specs would solve all this. It wouldn’t solve my past issue with the json schema validator, because it was `required: false` in the Gemfile.

Reply:
`require File.dirname(__FILE__) + "/../../spec_helper"` is some legacy, and can be written with a `require_relative`, `__dir__`, using the `/` on the `Pathname`, etc.

But we have to keep in mind that you can run specs from a directory inside the spec directory, not necessarily the project root.

And if there’s a `gems/` directory where you keep code that is to be eventually extracted, with their own `spec/` directory?

Reply:
We do not use rails anymore, as it was just to heavy back then. We require a separate require-structure, and it is not about bundler-requires, it is really about e.g. requiring the use-case class from within the worker-class or the api or whatever. 

And we sometimes had the issue that we started test-driven, implemented the spec, and obviously required the use-case-class there to be able to test it, and forgot to add it to the worker/api/.. then, but after deploying the api/worker failed due to missing require / unknown class. 

It did not happen often, but a few times already. So we thought: a easy rubocop would help us. And finally, after quite some time, we found the time to invest a bit of time here. 

I totally get to require stuff only where it is needed, as suggested in the comment you linked above, but imo this should only relate to "stuff needed for the test, and only the test", which tbh. we rarely have (and if, it is added to the spec_helper).

> I have no certainty that just banning requires from specs would solve all this. It wouldn’t solve my past issue with the json schema validator, because it was required: false in the Gemfile.

If it is not loaded via bundler, and not in your "config" somehow, the require in the spec-file would have covered it is not loaded on application start, so it might have solved your issue? Or did i get anything wrong?

Reply:
> only the test", which tbh. we rarely have

By “we”, do you mean the whole community, or your project?
Because above you’ve provided a “mostly” list that mostly has test-related dependencies.

I kind of agree that it’s a cleaner approach to use helpers. But I can’t recall if I’ve seen more than a handful.
Putting all into one or two helpers is not always the best option. Everyone hates when specs load the whole thing for half a minute. And every small dependency contributes to this.

At this point, I’m not convinced that requires from specs (except for helpers) are inherently bad, and should be banned.

> If it is not loaded via bundler

It was loaded by bundler, but only in test and development. In production, with `bundle --without development,test` there was no way it would be loaded.
I can’t think of anything that could have solved it.

Reply:
"We" is the projects I work on with others. One of them e.g. has this rspec run:

> Finished in 24.33 seconds (files took 7.11 seconds to load)
> 2989 examples, 0 failures

In this project we basically have everything either loaded in usual app code, or if only needed for specs, in spec_helper. Can't say it is slow tbh. ;-)

But nothing to discuss now. 

> It was loaded by bundler, but only in test and development. In production, with bundle --without development,test there was no way it would be loaded.
> I can’t think of anything that could have solved it.

True that, this would not have helped in that case. 

> At this point, I’m not convinced that requires from specs (except for helpers) are inherently bad, and should be banned.

I will discuss it a bit further with my team, and come back to you.

Reply:
Ok , i talked to my team and we aligned on closing this PR, we might still add it for us as we still think it is valuable, but i get why it will not yet added to the public repo. 

Thanks for your feedback.